[popover] Implement "check and possibly close popover stack" algorithm

rwlbuis opened the pull request:
#### f4a34ff850554d9c85d9114b14a3a12bb01f1ccf
<pre>
[popover] Implement &quot;check and possibly close popover stack&quot; algorithm
<a href="https://bugs.webkit.org/show_bug.cgi?id=254382">https://bugs.webkit.org/show_bug.cgi?id=254382</a>

Reviewed by Tim Nguyen.

Implement &quot;check and possibly close popover stack&quot; algorithm as specified here:
<a href="https://github.com/whatwg/html/pull/9048">https://github.com/whatwg/html/pull/9048</a>

This patch also imports the latest version of popover-target-element-disabled.html.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-target-element-disabled-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-target-element-disabled.html:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::checkAndPossiblyClosePopoverStackInternal):
* Source/WebCore/html/HTMLElement.h:
(WebCore::HTMLElement::checkAndPossiblyClosePopoverStack):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::removedFromAncestor):
(WebCore::HTMLFormControlElement::parseAttribute):
(WebCore::HTMLFormControlElement::disabledStateChanged):
(WebCore::HTMLFormControlElement::didChangeForm):
* Source/WebCore/html/HTMLFormControlElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):

Canonical link: <a href="https://commits.webkit.org/262440@main">https://commits.webkit.org/262440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b11dc7261bda540218c91b468345320105c9efa2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1525 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1555 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1608 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2444 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/1675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1506 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1617 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1614 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1538 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1380 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2282 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1393 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1364 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1389 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1403 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/2456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/1434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/1350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1370 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/381 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1487 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-target-element-disabled-expected.txt:
I think this test was updated to also check if popovertarget="" attribute changes also trigger this algorithm

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1284,6 +1284,23 @@ static HTMLElement* topmostPopoverAncestor(Element& newPopover)
     return topmostAncestor.get();
 }
 
+void HTMLElement::checkAndPossiblyClosePopoverStack()
```
Everywhere where we call this we'll want to check if the popovertarget attribute is present to avoid regressing speedometer. cc @cdumez

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1284,6 +1284,26 @@ static HTMLElement* topmostPopoverAncestor(Element& newPopover)
 
+void HTMLElement::checkAndPossiblyClosePopoverStack()
+{
+    if (!document().settings().popoverAttributeEnabled() || !document().hasTopLayerElement())
```
Can this check be inlined in the header to avoid a function call whenever the setting is disabled or the document has no top layer?

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1284,6 +1284,26 @@ static HTMLElement* topmostPopoverAncestor(Element& newPopover)
+    if (!document().settings().popoverAttributeEnabled() || !document().hasTopLayerElement())
+        return;
+
+    Vector<Element*> topLayerPositions;
```
Please don't store raw pointer in a container. Probably want to use WeakPtr or RefPtr here, depending on whether or not you'd like to keep the elements alive.

nt1m replied:
nit: can we name this variable `autoPopoverList` ?

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1284,6 +1284,26 @@ static HTMLElement* topmostPopoverAncestor(Element& newPopover)
+        topLayerPositions.append(element.ptr());
+    }
+
+    for (int i = topLayerPositions.size() - 1; i > 0; i--) {
```
size_t

cdumez commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -152,6 +153,8 @@ void HTMLFormControlElement::parseAttribute(const QualifiedName& name, const Ato
             m_isRequired = newRequired;
             requiredStateChanged();
         }
+    } else if (name == popovertargetAttr) {
```
No curly brackets for one-liners.

cdumez reviewed (commented):
Would be good to A/B test on speedometer

nt1m reviewed (commented):
Looks good with Chris' comments addressed, and with an A/B test after that. Thanks!

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1284,6 +1284,23 @@ static HTMLElement* topmostPopoverAncestor(Element& newPopover)
 
+void HTMLElement::checkAndPossiblyClosePopoverStackInternal()
+{
+    Vector<Element*> autoPopoverList;
```
This should be `Vector<RefPtr<Element>>` as Chris mentioned earlier

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1284,6 +1284,23 @@ static HTMLElement* topmostPopoverAncestor(Element& newPopover)
+    }
+
+    for (size_t i = autoPopoverList.size(); i > 1; i--) {
+        if (topmostPopoverAncestor(*autoPopoverList[i - 1]) != autoPopoverList[i - 2]) {
```
Can we write this to match the algorithm?

```suggestion
    for (size_t i = autoPopoverList.size() - 1; i > 0; i--) {
        if (topmostPopoverAncestor(*autoPopoverList[i]) != autoPopoverList[i - 1]) {
```

rwlbuis replied:
Yes, but then i needs to be an int since autoPopoverList.size() - 1 can be negative.

nt1m replied:
Alan suggested this pattern: `for (size_t i = autoPopoverList.size(); i--;)` which should keep `size_t`.

nt1m replied:
@rwlbuis can we follow this pattern?

```
for (size_t i = autoPopoverList.size(); i--;) {
    if (topmostPopoverAncestor(*autoPopoverList[i]) != autoPopoverList[i - 1]) {
```

rwlbuis replied:
That crashes, probably for the i == 0 case.

nt1m replied:
@alanbaradlay @cdumez wdyt?

nt1m replied:
@rwlbuis This should work:
```
for (size_t i = autoPopoverList.size(); i-- > 1;) {
    if (topmostPopoverAncestor(*autoPopoverList[i]) != autoPopoverList[i - 1]) {
```

rwlbuis replied:
That works locally, thanks!

nt1m reviewed (approved):
A/B testing shows this is fine. Thanks!

rwlbuis committed:
[popover] Implement "check and possibly close popover stack" algorithm
https://bugs.webkit.org/show_bug.cgi?id=254382

Reviewed by Tim Nguyen.

Implement "check and possibly close popover stack" algorithm as specified here:
https://github.com/whatwg/html/pull/9048

This patch also imports the latest version of popover-target-element-disabled.html.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-target-element-disabled-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-target-element-disabled.html:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::checkAndPossiblyClosePopoverStackInternal):
* Source/WebCore/html/HTMLElement.h:
(WebCore::HTMLElement::checkAndPossiblyClosePopoverStack):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::removedFromAncestor):
(WebCore::HTMLFormControlElement::parseAttribute):
(WebCore::HTMLFormControlElement::disabledStateChanged):
(WebCore::HTMLFormControlElement::didChangeForm):
* Source/WebCore/html/HTMLFormControlElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):

Canonical link: https://commits.webkit.org/262440@main
